Allow validators to detect transaction censorship attempts:

nbougalis opened the pull request:
The XRP Ledger is designed to be censorship resistant. Any attempt to censor transactions would require coordinated action by a majority of the system's validators.

Importantly, the design of the system is such that such an attempt is detectable and can be easily proven since every validators must sign the validations it publishes.

This commit adds an automated censorship detector which is active on servers configured as validators. The detector tracks all transactions proposed by the validator and issues warnings of increasing severity for any transactions which, despite being proposed repeatedly by the validator, has not been included.

ripplelabs-jenkins commented:
## Jenkins Build Summary

Built from [this commit](https://github.com/ripple/rippled/commit/9ef176c3164dcb7d1718e76c88d700d1b3db23c2)

Built at __20180930 - 03:07:34__

### Test Results

Build Type | Log | Result | Status
---------- | --- | ------ | ------
msvc.Debug | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2700/1//artifact/msvc.Debug.txt) | 1033 cases, 0 failed, t: 790s | PASS :white_check_mark: 
gcc.Debug <br/>-Dcoverage=ON | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2700/1//artifact/gcc.Debug_Dcoverage_ON.txt) | 1036 cases, 0 failed, t: 851s | PASS :white_check_mark: 
docs | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2700/1//artifact/docs.txt) | 1 cases, 0 failed, t: 2s | PASS :white_check_mark: 
gcc.Release <br/>-Dassert=ON, <br/>MANUAL_TESTS=true | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2700/1//artifact/gcc.Release_Dassert_ON_MANUAL_TESTS_true.txt) | 1101 cases, 0 failed, t: 933s | PASS :white_check_mark: 
clang.Debug | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2700/1//artifact/clang.Debug.txt) | 1036 cases, 0 failed, t: 284s | PASS :white_check_mark: 
rpm | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2700/1//artifact/rpm.txt) | 1035 cases, 0 failed, t: n/a | PASS :white_check_mark: 
clang.Debug <br/>-Dunity=OFF | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2700/1//artifact/clang.Debug_Dunity_OFF.txt) | 1036 cases, 0 failed, t: 441s | PASS :white_check_mark: 
gcc.Debug | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2700/1//artifact/gcc.Debug.txt) | 1036 cases, 0 failed, t: 290s | PASS :white_check_mark: 
msvc.Debug, <br/>NINJA_BUILD=true | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2700/1//artifact/msvc.Debug_NINJA_BUILD_true.txt) | 1033 cases, 0 failed, t: 651s | PASS :white_check_mark: 
gcc.Debug <br/>-Dunity=OFF | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2700/1//artifact/gcc.Debug_Dunity_OFF.txt) | 1036 cases, 0 failed, t: 439s | PASS :white_check_mark: 
clang.Release <br/>-Dassert=ON | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2700/1//artifact/clang.Release_Dassert_ON.txt) | 1035 cases, 0 failed, t: 405s | PASS :white_check_mark: 
gcc.Debug <br/>-Dstatic=OFF | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2700/1//artifact/gcc.Debug_Dstatic_OFF.txt) | 1036 cases, 0 failed, t: 292s | PASS :white_check_mark: 
gcc.Release <br/>-Dassert=ON | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2700/1//artifact/gcc.Release_Dassert_ON.txt) | 1035 cases, 0 failed, t: 423s | PASS :white_check_mark: 
gcc.Debug <br/>-Dstatic=OFF -DBUILD_SHARED_LIBS=ON | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2700/1//artifact/gcc.Debug_Dstatic_OFF_DBUILD_SHARED_LIBS_ON.txt) | 1036 cases, 0 failed, t: 292s | PASS :white_check_mark: 
gcc.Debug, <br/>NINJA_BUILD=true | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2700/1//artifact/gcc.Debug_NINJA_BUILD_true.txt) | 1036 cases, 0 failed, t: 275s | PASS :white_check_mark: 
msvc.Debug <br/>-Dunity=OFF | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2700/1//artifact/msvc.Debug_Dunity_OFF.txt) | 1033 cases, 0 failed, t: 1211s | PASS :white_check_mark: 
msvc.Release | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2700/1//artifact/msvc.Release.txt) | 1032 cases, 0 failed, t: 416s | PASS :white_check_mark: 
clang.Debug <br/>-Dunity=OFF -Dsan=address, <br/>PARALLEL_TESTS=false, <br/>DEBUGGER=false | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2700/1//artifact/clang.Debug_Dunity_OFF_Dsan_address_PARALLEL_TESTS_false_DEBUGGER_false.txt) | 1036 cases, 0 failed, t: 1068s | PASS :white_check_mark: 
clang.Debug <br/>-Dunity=OFF -Dsan=undefined, <br/>PARALLEL_TESTS=false | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2700/1//artifact/clang.Debug_Dunity_OFF_Dsan_undefined_PARALLEL_TESTS_false.txt) | 1036 cases, 0 failed, t: 1174s | PASS :white_check_mark:

bachase commented on src/ripple/app/ledger/impl/BuildLedger.cpp:
```diff
@@ -45,109 +45,97 @@ buildLedgerImpl(
+
             try
             {
+                if (pass == 0 && built->txExists(txid))
```
Is there a reason this filtering can't be done over in https://github.com/ripple/rippled/pull/2700/files#diff-256efa955bd129c0f8c25ae0f66d8afcR462 where `txns` is initially built?

bachase commented on src/ripple/app/consensus/RCLCensorshipDetector.h:
```diff
@@ -0,0 +1,128 @@
+                        It must return true for entries that should be removed.
+    */
+    template <class Predicate>
+    void prune(Predicate pred)
```
Make this private?

nbougalis replied:
This may be useful in different contexts (e.g. a command to display the contents of the detector) so I'd rather leave it.

bachase commented on src/ripple/app/consensus/RCLCensorshipDetector.h:
```diff
@@ -0,0 +1,128 @@
+namespace ripple {
+
+template <class TxID, class Sequence>
+class RCLCensorshipDetector
```
Consider adding unittests for this class.

nbougalis replied:
Wilco

nbougalis commented on src/ripple/app/misc/CanonicalTXSet.cpp:
```diff
@@ -106,28 +106,23 @@ CanonicalTXSet::prune(AccountID const& account,
+    auto tmp = it;
     ++tmp;
-    mMap.erase (it);
+    map_.erase (it);
```
Done.

MarkusTeufelberger commented:
Is is possible to do these checks on non-validating nodes too? Some "censorship probes" scattered throughout the network (to have better chances of receiving censored valid transactions) would probably be another use case for this and maybe even better suited for the task than validators.

nbougalis commented:
@MarkusTeufelberger: you _could_, I guess, but it could result in false positives, which is something I'd like to avoid. The protocol is such that if a validator proposes some transaction T that's valid but doesn't make it into consensus for round X, it's going to have priority for round X+1. That's something that doesn't happen with regular servers.

For that reason, I think it's better to only perform these checks on validators. If someone is interested in probing this, running a validator is no more expensive or computationally intensive than is running a regular node.

nbougalis commented:
I copied the header from some other file and didn’t really think much about it so, 2016. Will fix. Thanks!

JoelKatz commented:
I haven't reviewed yet, but two things stick out. First, typo in commit message, "every validator[s] must". Second, this probably should be enabled on non-validators as well as validators. All participants want to now if censorship is affecting the network.

seelabs commented on src/ripple/app/consensus/RCLCensorshipDetector.h:
```diff
@@ -0,0 +1,127 @@
+
+namespace ripple {
+
+template <class TxID, class Sequence>
```
It is much easier to reason about concrete classes than templates. It appears the reaons for making this a template is ease of testing - i.e. so we can use `int` instead of the more cumbersome `uint256`. I'd rather complicate the testing code (by say having an int->uint256 adaptor in the testing) than make this a template.

nbougalis replied:
We could, but the rest of the code is similarly templetized. This might be an interesting place to begin undoing that, I guess.

seelabs commented on src/ripple/app/consensus/RCLCensorshipDetector.h:
```diff
@@ -0,0 +1,127 @@
+class RCLCensorshipDetector
+{
+private:
+    std::map<TxID, Sequence> tracker;
```
Style nit: Since this is a private member, should be `tracker_`,

seelabs commented on src/ripple/app/consensus/RCLCensorshipDetector.h:
```diff
@@ -0,0 +1,127 @@
+                    It must return true for entries that should be removed.
+    */
+    template <class Predicate>
+    void prune(Predicate pred)
```
`Predicate` should be passed by forwarding ref (i.e. `prune(Predicate&& pred)`). Some of the predicates that are passed capture variables (including capturing vectros). These predicates are not cheap to copy.

seelabs commented on src/ripple/app/consensus/RCLCensorshipDetector.h:
```diff
@@ -0,0 +1,127 @@
+    */
+    void propose(
+        Sequence seq,
+        std::vector<TxID> proposed)
```
The reason `proposed` is passed by value is because of the `sort`. This complicated the interface because callers need to `std::move` the parameter to avoid the copy (if they don't use the collection again) or need to copy if they do - even if the collection was already sorted.

I honestly don't have a good solution. `boost::flat_set` is the best I can come up with. However, that has quadradic construction, and it's much more likely callers will already have a `std::vector`. Since N is small the extra construction doesn't bug me too much. If you don't like that (and I'm on the fence), then we at least need to change the callers to move this parameters. I will leave comments at the call sites.

seelabs commented on src/ripple/app/consensus/RCLCensorshipDetector.h:
```diff
@@ -0,0 +1,127 @@
+    */
+    template <class Predicate>
+    void check(
+        std::vector<TxID> accepted,
```
Same comments as above for passing `accepted` by value.

seelabs commented on src/ripple/app/consensus/RCLCensorshipDetector.h:
```diff
@@ -0,0 +1,127 @@
+    template <class Predicate>
+    void check(
+        std::vector<TxID> accepted,
+        Predicate pred)
```
Same comment as above: change `Predicate` to a forwarding ref

seelabs commented on src/ripple/app/consensus/RCLConsensus.cpp:
```diff
@@ -330,6 +329,19 @@ RCLConsensus::Adaptor::onClose(
+                proposed.push_back(item->key());
+            });
+
+        censorshipDetector_.propose(prevLedger->info().seq + 1, proposed);
```
Definately move here. Otherwise we'll allocate.

nbougalis replied:
:+1:

seelabs commented on src/ripple/app/consensus/RCLConsensus.cpp:
```diff
@@ -431,41 +443,104 @@ RCLConsensus::Adaptor::doAccept(
+        for (auto const& r : retriableTxs)
+            failed.insert (r.first.getTXID());
+
+        censorshipDetector_.check(accepted,
```
`accepted` should be `std::move` here

seelabs commented on src/ripple/app/consensus/RCLConsensus.h:
```diff
@@ -151,11 +160,11 @@ class RCLConsensus
             If not available, asynchronously acquires from the network.
 
-            @param ledger The ID/hash of the ledger acquire
+            @param hash The ID/hash of the ledger acquire
```
Not in this changeset, but on line 149 "called" is misspelled as "callled". On line 313 "believe" is misspelled as "believ".

seelabs commented on src/test/app/RCLCensorshipDetector_test.cpp:
```diff
@@ -0,0 +1,96 @@
+        // Finalize the round, by processing what we accepted; then
+        // remove anything that needs to be removed and ensure that
+        // what remains is correct.
+        cdet.check(accepted,
```
`std::move` accepted here.

seelabs commented on src/test/app/RCLCensorshipDetector_test.cpp:
```diff
@@ -0,0 +1,96 @@
+        std::vector<int> remain, std::vector<int> remove)
+    {
+        // Begin tracking what we're proposing this round
+        cdet.propose(round, proposed);
```
`std::move` proposed here.

nbougalis commented:
@JoelKatz: I considered making this available for non-validators, but I believe it would complicate things unnecessarily and not be as good. My rationale, copy-pasted from an earlier message to this thread:

>The protocol is such that if a validator proposes some transaction T that's valid but doesn't make it into consensus for round X, it's going to have priority for round X+1. That's something that doesn't happen with regular servers.
>
>For that reason, I think it's better to only perform these checks on validators. If someone is interested in probing this, running a validator is no more expensive or computationally intensive than is running a regular node.
>
>To do what you suggest would require adding more (and more complicated) code which is something I’d like to avoid.

If you believe it's super-important to have this operate on non-validators as well, we'll have to go back to the drawing board and design something more complicated.

nbougalis commented on src/ripple/app/consensus/RCLCensorshipDetector.h:
```diff
@@ -0,0 +1,128 @@
+                        It must return true for entries that should be removed.
+    */
+    template <class Predicate>
+    void prune(Predicate pred)
```
Unfortunately, `std::remove_if` doesn't work on associative containers like the `std::map` we're using.

JoelKatz commented:
I do feel fairly strongly that everyone who uses the system wants to know if they're seeing censorship, whether they're running a validator or not. If you see censorship, whether you're running a validator or not, that's unexpected and indicates something is wrong that a human should look at.

nbougalis commented:
@JoelKatz, you and @MarkusTeufelberger have convinced me that the best thing to do is to have this running on _every_ server and not just validators. I will update this PR accordingly. Thanks.

nbougalis commented:
Rebased against **1.2.0-b5** and modified the code so that every server performs censorship detection as long as it entered a round synced to the network.

nbougalis commented on src/ripple/app/consensus/RCLConsensus.cpp:
```diff
@@ -432,40 +452,97 @@ RCLConsensus::Adaptor::doAccept(
+
+    // Attempt to detect attempts at censorship of transaction which we propose
+    // by tracking which ones don't make it in after a period of time.
+    if (mode == ConsensusMode::proposing && result.state != ConsensusState::MovedOn)
```
Changed to `result.state == ConsensusState::Yes` along with other changes.

bachase commented on src/ripple/app/consensus/RCLConsensus.cpp:
```diff
@@ -735,6 +804,12 @@ RCLConsensus::Adaptor::onModeChange(
+
+    // If we were proposing but aren't any longer, we need to reset the
+    // censorship tracking to avoid bogus warnings.
+    if (before == ConsensusMode::proposing && before != after)
```
Now that both validators and non-validators are tracking, I believe this check also needs to change to handle when a non-validator server detects it is working on the wrong ledger:
```suggestion
    if ((before == ConsensusMode::proposing || before == ConsensusMode::observing) && before != after)
```

nbougalis commented:
That's an interesting point, and I agree that we should do this. But I think I'm going to punt right now. I want to get censorship detection merged so it can be included in **1.2.0**. We can augment the code to log in a future PR.
